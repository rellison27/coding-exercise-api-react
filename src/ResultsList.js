import React, { Component } from 'react'
import { Table } from 'semantic-ui-react'
import CSVReader from "react-csv-reader";

class ResultsList extends Component {
    constructor(props) {
        super(props);
        this.state = { data: [] };
    }

    componentDidMount() {
        fetch("http://localhost:8000/api/people")
          .then(response => response.json())
          .then(data => this.setState({ data: data.data }));
    }

    render() {
        var data = this.state.data || [];
        const handleForce = (data, fileInfo) => console.log(data, fileInfo);

        const papaparseOptions = {
          header: true,
          dynamicTyping: true,
          skipEmptyLines: true,
          transformHeader: header => header.toLowerCase().replace(/\W/g, "_")
        };
        // Step 3
        // Update the ReactJS application to 
        // receive an uploaded People & Group CSV file
        return (
          <>
          <CSVReader
      cssClass="react-csv-input"
      label="Select CSV with Groups or People"
      onFileLoaded={handleForce}
      parserOptions={papaparseOptions}
    />
            <Table celled padded>
              <Table.Header>
                <Table.Row>
                  <Table.HeaderCell singleLine>First Name</Table.HeaderCell>
                  <Table.HeaderCell>Last Name</Table.HeaderCell>
                  <Table.HeaderCell>Email</Table.HeaderCell>
                  <Table.HeaderCell>Status</Table.HeaderCell>
                </Table.Row>
              </Table.Header>

              <Table.Body>

              {
                  data.map((person, index) => {
                      return (
                          <Table.Row key={index}>
                              <Table.Cell singleLine>{ person.first_name }</Table.Cell>
                              <Table.Cell singleLine>{ person.last_name }</Table.Cell>
                              <Table.Cell singleLine>{ person.email_address }</Table.Cell>
                              <Table.Cell singleLine>{ person.status }</Table.Cell>
                          </Table.Row>
                      );
                    })
              }

              </Table.Body>
            </Table>
            </>
    );
}

}

export default ResultsList
