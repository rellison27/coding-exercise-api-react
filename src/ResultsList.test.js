import React from "react";
import { mount } from "enzyme";
import ResultsList from "./ResultsList";

let wrapper, data, people, groups;

describe("<ResultsList />", () => {
    beforeAll(() => {
        wrapper = mount(<ResultsList />);
        data = [
            {
                id: 132,
                first_name: "Macie",
                last_name: "Emmerich",
                email_address: "cremin.marjory@hotmail.com",
                status: "active",
                group_name: "Pastors",
                updated_at: "2019-07-20 22:05:47",
                created_at: "2019-07-20 22:05:47",
            },
            {
                id: 132,
                first_name: "Ellis",
                last_name: "Ferich",
                email_address: "bremin.marjory@hotmail.com",
                status: "archived",
                group_name: "Elders",
                updated_at: "2019-07-20 22:05:47",
                created_at: "2019-07-20 22:05:47",
            },
        ];

        wrapper.setState({ peopleData: data });
    });

    test("renders", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("Groups table shouldn't be present without GroupData", () => {
        //console.log(wrapper.find("Table").debug());
        expect(wrapper.find("Table")).toHaveLength(1);
    });

    test("People table should sort by first name", () => {
        expect(
            wrapper.find("TableBody TableRow TableCell").at(0).find("td").text()
        ).toContain("Macie");
        wrapper.find("Table TableHeaderCell").at(0).simulate("click");
        expect(
            wrapper.find("TableBody TableRow TableCell").at(0).find("td").text()
        ).toContain("Ellis");
    });

    test("People table should sort by last name", () => {
        expect(
            wrapper.find("TableBody TableRow TableCell").at(1).find("td").text()
        ).toContain("Ferich");
        wrapper.find("Table TableHeaderCell").at(1).simulate("click");
        expect(
            wrapper.find("TableBody TableRow TableCell").at(1).find("td").text()
        ).toContain("Emmerich");
    });

    test("People table should sort by email address", () => {
        expect(
            wrapper.find("TableBody TableRow TableCell").at(2).find("td").text()
        ).toContain("cremin.marjory@hotmail.com");
        wrapper.find("Table TableHeaderCell").at(2).simulate("click");
        expect(
            wrapper.find("TableBody TableRow TableCell").at(2).find("td").text()
        ).toContain("bremin.marjory@hotmail.com");
    });

    test("People table should sort by status", () => {
        expect(
            wrapper.find("TableBody TableRow TableCell").at(3).find("td").text()
        ).toContain("archived");
        wrapper.find("Table TableHeaderCell").at(3).simulate("click");
        expect(
            wrapper.find("TableBody TableRow TableCell").at(3).find("td").text()
        ).toContain("active");
    });

    test("People table should sort by group name", () => {
        expect(
            wrapper.find("TableBody TableRow TableCell").at(4).find("td").text()
        ).toContain("Pastors");
        wrapper.find("Table TableHeaderCell").at(4).simulate("click");
        expect(
            wrapper.find("TableBody TableRow TableCell").at(4).find("td").text()
        ).toContain("Elders");
    });
});

describe("<ResultsList /> with groups", () => {
    beforeAll(() => {
        wrapper = mount(<ResultsList />);
        people = [
            {
                id: 132,
                first_name: "Macie",
                last_name: "Emmerich",
                email_address: "cremin.marjory@hotmail.com",
                status: "active",
                group_name: "Elders",
                updated_at: "2019-07-20 22:05:47",
                created_at: "2019-07-20 22:05:47",
            },
            {
                id: 132,
                first_name: "Ellis",
                last_name: "Ferich",
                email_address: "bremin.marjory@hotmail.com",
                status: "active",
                group_name: "Elders",
                updated_at: "2019-07-20 22:05:47",
                created_at: "2019-07-20 22:05:47",
            },
        ];

        groups = [{ group_name: "Elders" }, { group_name: "Members" }];

        wrapper.setState({ peopleData: people, groupData: groups });
    });

    test("Two Group Tables should be present", () => {
        expect(wrapper.find("Table")).toHaveLength(3);
    });

    test("First Table should be Elders", () => {
        expect(
            wrapper.find("Table").at(1).find("TableHeaderCell").text()
        ).toContain("Elders");
    });
    test("Second Table should be Members", () => {
        expect(
            wrapper.find("Table").at(2).find("TableHeaderCell").text()
        ).toContain("Members");
    });
});
