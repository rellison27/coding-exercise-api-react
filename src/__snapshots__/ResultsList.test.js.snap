// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsList /> renders 1`] = `
<ResultsList>
  <div>
    Upload a .CSV file for your Groups and/or People
  </div>
  <a
    cssClass="react-csv-input"
    onFileLoaded={[Function]}
    parserOptions={
      Object {
        "dynamicTyping": true,
        "header": true,
        "skipEmptyLines": true,
        "transformHeader": [Function],
      }
    }
  >
    <div
      className="react-csv-input"
    >
      <input
        accept=".csv, text/csv"
        className="csv-input"
        disabled={false}
        id="react-csv-reader-input"
        onChange={[Function]}
        style={Object {}}
        type="file"
      />
    </div>
  </a>
  <h1>
     People
  </h1>
  <Table
    as="table"
    celled={true}
    padded={true}
    sortable={true}
  >
    <table
      className="ui celled sortable padded table"
    >
      <TableHeader
        as="thead"
      >
        <thead
          className=""
        >
          <TableRow
            as="tr"
            cellAs="td"
          >
            <tr
              className=""
            >
              <TableHeaderCell
                as="th"
                onClick={[Function]}
                singleLine={true}
                sorted={null}
              >
                <TableCell
                  as="th"
                  className=""
                  onClick={[Function]}
                  singleLine={true}
                >
                  <th
                    className="single line"
                    onClick={[Function]}
                  >
                    First Name
                  </th>
                </TableCell>
              </TableHeaderCell>
              <TableHeaderCell
                as="th"
                onClick={[Function]}
                sorted={null}
              >
                <TableCell
                  as="th"
                  className=""
                  onClick={[Function]}
                >
                  <th
                    className=""
                    onClick={[Function]}
                  >
                    Last Name
                  </th>
                </TableCell>
              </TableHeaderCell>
              <TableHeaderCell
                as="th"
                onClick={[Function]}
                sorted={null}
              >
                <TableCell
                  as="th"
                  className=""
                  onClick={[Function]}
                >
                  <th
                    className=""
                    onClick={[Function]}
                  >
                    Email Address
                  </th>
                </TableCell>
              </TableHeaderCell>
              <TableHeaderCell
                as="th"
                onClick={[Function]}
                sorted={null}
              >
                <TableCell
                  as="th"
                  className=""
                  onClick={[Function]}
                >
                  <th
                    className=""
                    onClick={[Function]}
                  >
                    Status
                  </th>
                </TableCell>
              </TableHeaderCell>
              <TableHeaderCell
                as="th"
                onClick={[Function]}
                sorted={null}
              >
                <TableCell
                  as="th"
                  className=""
                  onClick={[Function]}
                >
                  <th
                    className=""
                    onClick={[Function]}
                  >
                    Group Name (Optional)
                  </th>
                </TableCell>
              </TableHeaderCell>
            </tr>
          </TableRow>
        </thead>
      </TableHeader>
      <TableBody
        as="tbody"
      >
        <tbody
          className=""
        >
          <TableRow
            as="tr"
            cellAs="td"
            key="0"
          >
            <tr
              className=""
            >
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  Macie
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  Emmerich
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  cremin.marjory@hotmail.com
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  active
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  Pastors
                </td>
              </TableCell>
            </tr>
          </TableRow>
          <TableRow
            as="tr"
            cellAs="td"
            key="1"
          >
            <tr
              className=""
            >
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  Ellis
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  Ferich
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  bremin.marjory@hotmail.com
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  archived
                </td>
              </TableCell>
              <TableCell
                as="td"
                singleLine={true}
              >
                <td
                  className="single line"
                >
                  Elders
                </td>
              </TableCell>
            </tr>
          </TableRow>
        </tbody>
      </TableBody>
    </table>
  </Table>
  <h1>
    Groups
  </h1>
</ResultsList>
`;
